### DISCUSSION TITLE
Is there a recommended way to remove temporary directories and files?

### DISCUSSION THREAD

**Discussion by milotheirself** [denoland/deno#14715] [Q&A] [answered]
Hi, I was wondering when and how it is sensible to manually remove temporary directories that are only required while a process is running. Initially, I was using the unload event to remove all directories after the process had successfully completed –

```ts
const tempPath = await Deno.makeTempDir();

globalThis.onunload = (e: Event): void => {
  Deno.remove(tempPath , { recursive: true });
};
```

However, the function is never called and the directory remains when the process terminates due to an exception.  So, for the time being, I'm removing any old directories from previous runs at the start of the process.  Similar to this – 

```ts
const tempPath = await Deno.makeTempDir({ prefix: `my-prefix-` });

for await (const dirEntry of Deno.readDir(path.resolve(tempPath, `../`))) {
  if (!dirEntry.name.startsWith(`my-prefix-`)) continue;
 
  const olderPath = path.resolve(tempPath, `../${dirEntry.name}`);

  if (tempPath != olderPath) {
    Deno.remove(olderPath, { recursive: true });
  }
}
```

Are there recommended methods for doing this with Deno?

**Comment by nayeemrmn** [comment, marked as answer]
Normally you should remove them right after using them in a `finally` block:
```ts
const tempPath = await Deno.makeTempDir({ prefix: `my-prefix-` });
try {
  // code that uses `tempPath`...
} finally {
  await Deno.remove(tempPath, { recursive: true });
}
```

**Comment by cjihrig** [comment]
> However, the function is never called and the directory remains when the process terminates due to an exception.

What exception are you seeing? Your first code snippet runs on my machine.

**Reply by milotheirself** [reply]
I meant exceptions from other parts of my code while developing. ^^

**Reply by cjihrig** [reply]
Ah got it. Using something like a `finally` block sounds like a good idea as @nayeemrmn mentioned. However, you still might want to consider doing the clean up at the start of the process anyway because you never really know what state the file system might be in when you start up.
